### PR TITLE
Adding PSI from scratch

### LINKED ISSUE #5: e2e working flow for PSI using websockets
Functional PSI using websockets

### LINKED ISSUE #6: Integrate PSI into zk summit app
Working PSI connected to user collected signatures

### PR DESCRIPTION
**Pull request by RiverRuby**
Closes #6
Closes #5 

UX is incredible. Supabase Realtime was just what we needed to make this clean. I'm very, very hyped. We are pushing the boundaries here

### CODE REVIEW

File: src/lib/client/jubSignalClient.ts [hunk 1]
```diff
@@ -219,6 +244,7 @@ const processEncryptedMessages = async (args: {
               name: metadata.fromDisplayName,
               encPk: metadata.fromPublicKey,
               sigPk: pk,
+              pkId: "0",
```
Comment (AndrewCLu) [inline src/lib/client/jubSignalClient.ts:247]: How does this work again? My id isn't actually 0 but i set it to 0 so i'm never included in the overlap?

File: src/lib/client/jubSignal/inboundTap.ts [hunk 2]
```diff
@@ -19,13 +21,17 @@ export const inboundTapMessageSchema = object({
...
+  pkId: string().required(),
+  psiPkLink: string().required(),
   msg: string().required(),
   sig: string().required(),
 });
 
 export type EncryptInboundTapMessageArgs = {
   displayName: string;
   encryptionPublicKey: string;
+  psiPublicKeysLink: string;
```
Comment (AndrewCLu) [inline src/lib/client/jubSignal/inboundTap.ts:33]: Grr...you changed the ordering of these two vars :(

File: src/lib/client/localStorage/keys.ts [hunk 3]
```diff
@@ -10,20 +10,31 @@ export const KEYS_STORAGE_KEY = "keys";
...
+    signaturePrivateKey,
+    psiPrivateKeys,
+    psiPublicKeysLink,
+  } = keys;
   const currentKeys = getFromLocalStorage(KEYS_STORAGE_KEY);
   if (currentKeys) {
     const parsedKeys = JSON.parse(currentKeys);
     parsedKeys.encryptionPrivateKey = encryptionPrivateKey;
     parsedKeys.signaturePrivateKey = signaturePrivateKey;
+    parsedKeys.psiPrivateKeys = psiPrivateKeys;
```
Comment (AndrewCLu) [inline src/lib/client/localStorage/keys.ts:36]: How large is psePrivateKeys?

File: src/lib/client/localStorage/users.ts [hunk 4]
```diff
@@ -36,65 +39,6 @@ export const getUsers = (): Record<string, User> => {
   return {};
 };
 
-// Populate user information based on a tap
-export const updateUserFromTap = async (
```
Comment (AndrewCLu) [inline src/lib/client/localStorage/users.ts:40]: Thanks! Missed this

File: src/lib/client/psi.ts [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    psiPublicKeys: gen_keys_output.message_round1,
+  };
+};
+
+export const generateSelfBitVector = (): Uint32Array => {
+  let bitVector = new Uint32Array(1000).fill(0);
+  let users = getUsers();
+  let locations = getLocationSignatures();
+
+  // 0-500 reserved for users
```
Comment (AndrewCLu) [inline src/lib/client/psi.ts:30]: If we are pregenerating keys for everything, can we just hardcode the map of public keys to indices somewhere?

File: src/lib/mp_psi/LICENSE [hunk 6]
```diff
@@ -0,0 +1,21 @@
+MIT License
+
+Copyright (c) 2024 Gaussian
```
Comment (AndrewCLu) [inline src/lib/mp_psi/LICENSE:3]: This is why I thought they were called Gaussian

File: src/lib/shared/signature.ts [hunk 7]
```diff
@@ -1,5 +1,5 @@
 import {
-  generateSignatureKeyPair as gebBBJJKeyPair,
```
Comment (AndrewCLu) [inline src/lib/shared/signature.ts:2]: was this here the whole time lmfao

File: src/lib/client/psi.ts [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+    if (user.pkId && user.pkId !== "0") {
+      let index = parseInt(user.pkId);
+      if (index < 500) {
+        bitVector[index] = 1;
+      }
+    }
+  }
+
+  // 500-1000 reserved for locations
+  for (let id in locations) {
```
Comment (AndrewCLu) [inline src/lib/client/psi.ts:43]: How exactly are the location and user indices decided? Where are they set?

File: src/pages/login.tsx [hunk 9]
```diff
@@ -159,6 +160,15 @@ export default function Login() {
       });
     }
 
+    const { data: authData, error: authError } =
+      await supabase.auth.signInAnonymously();
```
Comment (AndrewCLu) [inline src/pages/login.tsx:164]: How does this work? How does supabase know what permission levels the user has? Doesn't seem like we are providing a token or anything

File: src/pages/users/[id]/index.tsx [hunk 10]
```diff
@@ -32,19 +44,281 @@ const LinkCard = ({ label, value, href }: LinkCardProps) => {
...
+          });
+        }
+      });
+  };
+
+  // process broadcast events
+  useEffect(() => {
+    if (!broadcastEvent) return;
+
+    console.log(broadcastEvent);
```
Comment (AndrewCLu) [inline src/pages/users/[id]/index.tsx:170]: Oops

File: src/pages/users/[id]/index.tsx [hunk 11]
```diff
@@ -32,19 +44,281 @@ const LinkCard = ({ label, value, href }: LinkCardProps) => {
...
+
+    const { payload } = broadcastEvent;
+    if (payload.state === PSIState.ROUND2 && payload.to === selfEncPk) {
+      setOtherRound2MessageLink(payload.data);
+      console.log(parseInt(payload.otherPkId), parseInt(user?.pkId!));
```
Comment (AndrewCLu) [inline src/pages/users/[id]/index.tsx]: Oops

File: src/pages/users/[id]/index.tsx [hunk 12]
```diff
@@ -32,19 +44,281 @@ const LinkCard = ({ label, value, href }: LinkCardProps) => {
...
+  const [otherRound2MessageLink, setOtherRound2MessageLink] =
+    useState<string>();
+  const [selfRound2Output, setSelfRound2Output] = useState<any>();
+  const [round2Order, setRound2Order] = useState<boolean>();
+  const [otherRound3MessageLink, setOtherRound3MessageLink] =
+    useState<string>();
+  const [selfRound3Output, setSelfRound3Output] = useState<any>();
+
+  // set up channel for PSI
+  const setupChannel = () => {
```
Comment (AndrewCLu) [inline src/pages/users/[id]/index.tsx:88]: This is CLEAN

**Review by AndrewCLu — APPROVED** [review]
Amazing!! Great abstractions, really nice to see how real time database makes this so much easier. Just left a few questions and small comments, but this looks good to go.

File: src/lib/client/jubSignalClient.ts (hunk 1 above)
Comment (RiverRuby) [reply src/lib/client/jubSignalClient.ts:247]: Yeah, it's a bit of a hack

File: src/lib/client/localStorage/keys.ts (hunk 3 above)
Comment (RiverRuby) [reply src/lib/client/localStorage/keys.ts:36]: very small

Comment (RiverRuby) [reply src/lib/client/localStorage/keys.ts:36]: 0.0030221939086914062 MB

File: src/lib/shared/signature.ts (hunk 7 above)
Comment (RiverRuby) [reply src/lib/shared/signature.ts:2]: lmfao YEP

File: src/lib/client/psi.ts (hunk 8 above)
Comment (RiverRuby) [reply src/lib/client/psi.ts:43]: just from the ids in DB, which are autoincrement

File: src/pages/login.tsx (hunk 9 above)
Comment (RiverRuby) [reply src/pages/login.tsx:164]: supabase handles it by storing in localStorage. we don't need any specific permissions to do PSI stuff

File: src/lib/client/psi.ts (hunk 5 above)
Comment (RiverRuby) [reply src/lib/client/psi.ts:30]: we can, but the current method works fine!

**Pull request merged**